modules/post/windows/gather/enum_ms_product_keys.rb

```diff
@@ -120,11 +126,16 @@ def decode(chunk)
       end
     end
 
-    return key.reverse
+    key.reverse
   end
 
   def run
-    print_status("Finding Microsoft key on #{sysinfo['Computer']}")
+    if !sysinfo.nil? && sysinfo['Architecture'] == ARCH_X64 && session.arch == ARCH_X86
+      fail_with(Failure::NoTarget, 'Running via WOW64 is not supported, try using an x64 meterpreter...')
```

Comment:
Would this work on WOW64 if the [view](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/windows/registry.rb#L19) were set to the appropriate value when calling `registry_getvaldata` on L63-68 (and others if there are more that I missed).

Reply:
Maybe. The optional `view` parameter is rarely used in framework. Same with WOW64 handling/support.

```
# grep -rn 'registry_get' modules/ lib | grep VIEW 
modules/post/windows/gather/credentials/epo_sql.rb:51:    dbprops_file = registry_getvaldata(epol_reg_key, 'TomcatFolder', REGISTRY_VIEW_32_BIT)
modules/post/windows/manage/rollback_defender_signatures.rb:62:    disable_key_value = meterpreter_registry_getvalinfo(defender_disable_key, "DisableAntiSpyware", REGISTRY_VIEW_NATIVE)
lib/msf/core/post/windows/registry.rb:185:  def registry_getvaldata(key, valname, view = REGISTRY_VIEW_NATIVE)
lib/msf/core/post/windows/registry.rb:196:  def registry_getvalinfo(key, valname, view = REGISTRY_VIEW_NATIVE)
```

Reply:
On x86_64 Windows 7 / 2008 hosts:

* `REGISTRY_VIEW_NATIVE` did not work on shell or meterpreter.
* `REGISTRY_VIEW_32_BIT` did not work on shell or meterpreter.
* `REGISTRY_VIEW_64_BIT` worked on shell and meterpreter.

On x86 Windows 7 host:

* `REGISTRY_VIEW_NATIVE` worked on shell and meterpreter.
* `REGISTRY_VIEW_32_BIT` worked on meterpreter but not on shell.
* `REGISTRY_VIEW_64_BIT` worked on meterpreter but not on shell.

Reply:
It appears that if the host is x86_64 and the session is x86 meterpreter, then using `REGISTRY_VIEW_64_BIT` offers an improvement. In every other instance, doing nothing (providing no `view`) is as good as it gets (without modifying libraries).

Ideally this would be resolved in a library or at least using consistent wow64 detection copypasta between modules. I believe we only have copypasta for wow64 detection on Meterpreter sessions, not shell sessions.

I've updated the PR to support Meterpreter WOW64. The module will fail with no error message for WOW64 on shell sessions. This is a systemic issue across the entirety of Framework. WOW64 support is poor.